### PR TITLE
Reworked variables collectors

### LINKED ISSUE #163: Refactoring ideas
- [x] Create `Form` and `FormField` value objects in `Tempalte` namespace - do not use `Collected*` versions in template/compiler (Collected versions are between collector and finder, normal versions are between finder and template/compiler)
- [x] Create `TemplateContext` value object that would group variables, components, forms and filters passed to tempalte in single object
- [x] ~~`Template` value object might be renamed to `TemplateRender` becaause it prepresents specific render of template~~
- [x] `Tempalte` object should internally use new `TemplateContext`
- [x] ~~Create `TemplateContextFinder` thet would return `TempalteContext` for given class/method - then resolvers would use mainly this finder and would not need to change whenever we add something new to `TempalteContext` (like we did with Forms and Filters recently)~~
- [ ] Remove `VariableCollector` namespace completely
  * `ʟ_*` variables should be handled by `Compiler`
  * `presenter`, `control`, `flashes` should be handled by `LatteTempalteResolver` (on things that actually sets them like presenters/controls)
  * not sure about `baseUrl` and `basePath` - we should investigate where they are set in Latte
  * dynamic filter variables should be handled by `Compiler`
  * globalTemplateVariables could be dedicated NodeVisitor that takes them as parameter
  * Maybe create something like `AbstractVarTypeNodeVisitor` that would contain logic of adding variables. New variables then could be added by creating additional NodeVisitors
- [x] Create `ComponentsHelper` with merge/union logis similar to `VariablesHelper` to properly handle conflicting definitions of component with same name (for example in parent and child)
- [x] Maybe change latte context resolvers to support multiple nodes at once like rector? To avoid unnecessary calls on irrelevant nodes.
- [x] Create LatteContextCollectorRegistry top handle logic of searching for relevant collectors for given node. Similar to PHPStan\Collectors\Registry
- [ ] (???) Create CollectLocaltion value object to hold class+method where thing was colected
- [ ] (???) Maybe create common ancestor for value objects that have class/method context to avoid duplication

### PR DESCRIPTION
**Pull request by MartinMystikJonas**
Resolves #163

### CODE REVIEW

File: extension.neon [hunk 1]
```diff
@@ -116,10 +116,10 @@ services:
             - addNodeVisitor(50, Efabrica\PHPStanLatte\Compiler\NodeVisitor\AddParametersForBlockNodeVisitor())
             - addNodeVisitor(50, Efabrica\PHPStanLatte\Compiler\NodeVisitor\CopyDefinedVarsToOtherMethodsNodeVisitor())
             - addNodeVisitor(100, Efabrica\PHPStanLatte\Compiler\NodeVisitor\AddTypeToComponentNodeVisitor())
-            - addNodeVisitor(100, Efabrica\PHPStanLatte\Compiler\NodeVisitor\AddVarTypesNodeVisitor())
+            - addNodeVisitor(100, Efabrica\PHPStanLatte\Compiler\NodeVisitor\AddVarTypesNodeVisitor(%latte.globalVariables%))
             - addNodeVisitor(100, Efabrica\PHPStanLatte\Compiler\NodeVisitor\RemoveExtractParamsNodeVisitor())
             - addNodeVisitor(100, Efabrica\PHPStanLatte\Compiler\NodeVisitor\RenderBlockNodeVisitor())
-            - addNodeVisitor(200, Efabrica\PHPStanLatte\Compiler\NodeVisitor\ChangeFiltersNodeVisitor())
+            - addNodeVisitor(200, Efabrica\PHPStanLatte\Compiler\NodeVisitor\ChangeFiltersNodeVisitor(...))
```
Comment (lulco) [inline extension.neon:122]: What does `…` mean here?

**Review by lulco — APPROVED** [review]
I think this PR could be omitted and focus on refactoring with https://github.com/efabrica-team/phpstan-latte/issues/143 where each Template has its own engine and it means there are no global variables, no filters, no macros etc. Everything depends on which engine is used. Also each nette template class (DefaultTemplate as default) has its own variables.
But if you already implement it… it's up to you if you will merge it. Just saying that this code will be reworked again sooner or later.

**Comment by MartinMystikJonas** [comment]
I will merge it. When we willngo for customisable engines something similar to this would be needed anyway. Only some vars that are hardcoded now would be loaded based on engine/template in vartypesnodevisitor.

File: extension.neon (hunk 1 above)
Comment (MartinMystikJonas) [reply extension.neon:122]: Autowire all constructor params

Comment (lulco) [reply extension.neon:122]: That's default, isn't it?

Comment (MartinMystikJonas) [reply extension.neon:122]: Seems like it do not work implicitly inside setups. I had to add it to prevent error

Comment (lulco) [reply extension.neon:122]: Interesting. We have 2 params in https://github.com/efabrica-team/phpstan-latte/blob/main/src/Compiler/NodeVisitor/LinkNodeVisitor.php and it works from setup

Comment (MartinMystikJonas) [reply extension.neon:122]: 🤔🤷

**Pull request merged**